Add Support for Searching Processing Timelines

jbaptperez opened the pull request:
**IMPORTANT: All Pull Requests should be connected to an issue, if you don't
have an issue, please start by creating an issue and link it to the PR.**

Please provide enough information so that others can review your pull request:

<!-- You can skip this if you're fixing a typo or adding a minor bug fix -->

+ What existing problem does this PR solve?
  - Impossible to get timeline events when the one is being updated (new Plaso ingestion for the timeline).
+ What new feature is being introduced with this PR?
  - Allows to get timeline events when the timeline is being updated (new Plaso ingestion) i.e. allows to get partial events, already indexed,
  - Adds a setting to enable or disable this feature,
  - The UI clearly communicates to the user that search results against processing timelines may be incomplete and subject to change.
+ Overview of changes to existing functions if required.
  - Backend: In the `Sketch` model, the `active_timelines` property includes timelines with a status set to `processing`,
  - Backend: In the `utils.py`, the `get_validated_indices` function includes timelines with a status set to `processing`,
  - Backend: The `POST /sketches/{{sketch_id}}/aggregation/explore/` endpoint includes indices which timeline has a status set to `processing`.

**Checks**

- [x] All tests succeed.
- [ ] Unit tests added.
- [ ] e2e tests added.
- [ ] Documentation updated.

**Closing issues**

Closes #3219.

jkppr commented:
Thanks for adding a draft PR for early feedback. If I see this correctly, you are also having the changes submitted in your PR #3226 in this PR. Can you please separate this out and only commit the changes relevant to the issue #3219 in this PR? 

This would make reviewing the code much easier, thanks.

jbaptperez commented:
@jkppr, I removed the work of the other PR.
I also rebased the branch onto `master`.

The only change I made for now is at the backend side: It unlocks the desired behaviour: Getting timeline events while in the _processing_ status.
Note for now I did not added any frontend or backend setting to properly toggle this feature, I'll work on it at the end.

At the frontend side, for now, all timeline events appear even if some of them are is in the _processing_ state.
It is easy to reproduce such a situation by forcing a timeline state in the database.
This results in an infinite spinning wheel in the timeline chip.

**However**, I face difficulties when attempting to change the timeline chip so that a _processing_ one behaves like a _ready_/_failed_ one, with an additional spinning wheel as the only visual difference.

The complementarity between `TimelineChip.vue` and `TimelineComponent.vue` is not trivial.

Do you think someone can help me to accomplish this precise part?

jbaptperez commented:
Hi @jkppr, I made progress, but still WIP.

For now:

* I added a user setting (false by default): *showProcessingTimelineEvents*,
* I adapted the _frontend-ng_ so that it properly reads the setting,
* The frontend can display _processing_ timeline events with a consistent TimelineComponent/TimelineChip set,
* I have manually checked that there is no issue when displaying the timeline chip in the_Analyser results_ and the _Visualizations_ pages.

I am currently developing the following:

* Updating the backend so that the current hard-coded changes in my branch become variable, depending on a default value as a method parameter or coming from the related endpoint forms (at least `GET /sketches/{sketch_id}/`),
* Updating the frontend (_store.js_, _RestApiClient.js_) so that the setting is properly dispatched over the views.

Feel free to give me advice or instructions to fine tune my change, if necessary.

jbaptperez commented:
Hi @jkppr, I just pushed a first version of the feature.

Changes are:

* `POST /sketch/{sketchId}/explore`: Add an optional _includeProcessingTimelines_ parameter,
* _sketch_ model: Update the `active_timelines` property to include timelines with status _processing_,
* `POST /sketches/{sketch_id}/aggregation/explore/` : Include indices of timelines with status _processing_,
* Add a frontend user setting: _showProcessingTimelineEvents_,
* Update the frontend to take into account the _showProcessingTimelineEvents_,
* Add a frontend banner when displaying _processing_ timeline events (at least one processing timeline must be selected).

I will try to deploy those changes in my company in order to let the PO check the feature.
However, **I would like your opinion about the code I produced**, in the meantime.

I'll update the current code in the case we find a bug or the PO wants to change something.
I'll pull out the draft status of the PR when we'll consider it as final.

jkppr reviewed (changes requested):
This feature turned out great already during testing, thanks for all the effort!

Here is the first part of a partial review. It covers all files except some files in `components/Explore/`.

Some general topics I noticed:
* With the current implementation the decision if the feature will be used is solely on the user. What I'm missing is an option for a server admin to control the functionality. So adding a config option to `data/timesketch.conf` that allows admins to disable/enable the feature for their instance. Check server side functionality against that setting and handle accordingly. This does not only give the admin of an instance the decision what features they want to support on the deployment but also allows us later to easily add default settings for all users.
  * E.g. `SEARCH_PROCESSING_TIMELINES = True`
* There seems to be no processing progress shown anymore on the Data Source dialog, when the feature is turned on. Is this on purpose?

Otherwise I just noticed some errors and added suggestions or questions below.

jbaptperez commented on timesketch/models/sketch.py:
```diff
@@ -144,7 +144,7 @@ def active_timelines(self):
             timeline_status = timeline.get_status.status
             index_status = timeline.searchindex.get_status.status
-            if (timeline_status or index_status) in ("processing", "fail", "archived"):
+            if (timeline_status or index_status) in ("fail", "archived"):
```
Done (rebased).

jbaptperez commented on timesketch/lib/utils.py:
```diff
@@ -526,21 +526,27 @@ def read_and_validate_jsonl(
     """
+    allowed_statuses = ["ready"]
+    if include_processing_timelines:
+        allowed_statuses.append("processing")
```
Done (rebased).

jbaptperez commented on timesketch/frontend-ng/src/components/Explore/EventActionMenu.vue:
```diff
@@ -59,13 +59,16 @@ export default {
     },
     copyEventAsJSON() {
-      ApiClient.getEvent(this.sketch.id, this.event._index, this.event._id)
+      ApiClient.getEvent(this.sketch.id, this.event._index, !!this.settings.includeProcessingTimelines)
```
Fixed (rebased).

jbaptperez commented on timesketch/frontend-ng/src/components/Visualization/VisualizationEditor.vue:
```diff
@@ -438,7 +441,8 @@ export default {
           },
-        }
+        },
+        include_processing_timelines: !!this.settings.showProcessingTimelineEvents,
```
It adds the "processing yes/no" boolean for a further backend API call (to propagate the user setting up to the API call).

jbaptperez commented on timesketch/frontend-ng/src/components/SettingsDialog.vue:
```diff
@@ -76,6 +76,17 @@ limitations under the License.
+          >Select to include events from timelines in <strong>processing</strong> state</v-list-item-subtitle
+          >
+        </v-list-item-content>
+      </v-list-item>
```
Done (rebased).

jbaptperez commented on timesketch/frontend-ng/src/components/Visualization/SavedVisualization.vue:
I forgot that part. Implementation done (rebased).

jbaptperez commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -624,7 +626,8 @@ export default {
           end_time: endTime.toISOString().slice(0, -1),
-        }
+        },
+        include_processing_timelines: !!this.settings.showProcessingTimelineEvents,
```
Done (rebased).

jbaptperez commented on timesketch/frontend-ng/src/components/Explore/EventList.vue:
```diff
@@ -1107,6 +1123,15 @@ export default {
+      // least one enabled timeline is the "processing" state.
+      this.showBanner =
+        !!this.settings.showProcessingTimelineEvents &&
+        this.$store.state.sketch.active_timelines
```
Done (rebased).

jbaptperez commented on timesketch/frontend-ng/src/store.js:
```diff
@@ -292,6 +293,7 @@ export default new Vuex.Store({
           field: 'data_type',
           limit: '1000',
         },
+        include_processing_timelines: !!state.settings.showProcessingTimelineEvents,
```
Fixed (rebased).

jbaptperez commented on timesketch/lib/utils.py:
```diff
@@ -526,21 +526,27 @@ def read_and_validate_jsonl(
         indices: List of indices from the user
         sketch: A sketch object (instance of models.sketch.Sketch).
+        include_processing_timelines: True to include processing
+          timelines, Flase otherwise.
```
Done (rebased).

jbaptperez commented on timesketch/frontend-ng/src/components/Visualization/SavedVisualization.vue:
```diff
@@ -80,7 +83,8 @@ export default {
           this.parameters = JSON.parse(agg.parameters)
-          
+          this.parameters['include_processing_timelines'] = !!this.settings.showProcessingTimelineEvents
+
```
Code residue. Deleted (rebased).

jbaptperez commented:
@jkppr, @Annoraaq, we (a colleague and I) did the requested changes.
We rebased/squashed the result to minimize the amount of useless commits.
We only put the addition of the new system (backend) setting into a new commit.

The PR is now ready for a final review.

jkppr commented on timesketch/frontend-ng/src/components/LeftPanel/TimelinesTable.vue:
Small UI issue on the left panel:
![image](https://github.com/user-attachments/assets/f8b83153-5182-479f-94ec-a8f0ebab8469)

I think we can clip away a bit from the truncated text here if the timeline is in processing state.

jbaptperez replied:
Fixed (new commit).

jkppr commented on timesketch/frontend-ng/src/views/Analyze.vue:
```diff
@@ -17,7 +17,7 @@ limitations under the License.
     <v-card flat class="pa-3 pt-0 mt-n3" color="transparent">
       <div class="mt-2">
-        <ts-timeline-search componentName="analysis" :analyzer-timeline-id="analyzerTimelineId" @selectedTimelines="timelineSelection = $event"></ts-timeline-search>
+        <ts-timeline-search componentName="analysis" :includeProcessingTimelines="false" :analyzer-timeline-id="analyzerTimelineId" @selectedTimelines="timelineSelection = $event"></ts-timeline-search>
```
Like with the other comments regarding the analyzer code, let's remove it for this PR.

```suggestion
        <ts-timeline-search componentName="analysis" :analyzer-timeline-id="analyzerTimelineId" @selectedTimelines="timelineSelection = $event"></ts-timeline-search>
```

jbaptperez replied:
Same thing here, this is an explicit property value for the component (avoids "undefined").

jkppr reviewed (changes requested):
Some last minor changes.

jbaptperez commented on timesketch/frontend-ng/src/components/Analyzer/TimelineChip.vue:
```diff
@@ -38,6 +38,9 @@ limitations under the License.
             </template>
             <span>{{ timeline.name }}</span>
           </v-tooltip>
+          <span v-if="timeline.status[0].status === 'processing'" class="ml-3 mr-3">
+            <v-progress-circular small indeterminate color="grey" :size="17" :width="2"></v-progress-circular>
+          </span>
```
Actually, the `src/components/Analyzer/TimelineChip.vue` component is used in 2 places:

* When selecting timelines to run analyzers,
* When selecting timelines for a new visualization.

Run analyzers:
![image](https://github.com/user-attachments/assets/aa6dae28-a107-4e8b-91c2-d3adade51ef7)

New visualization:
![image](https://github.com/user-attachments/assets/b55784ae-b08e-432d-9501-a59aeb8454f0)

In the first case, we intentionally prevented the selection of processing timemines: They don't appear.
Otherwise, in the second case, a visualization is possible, so the chip should appear, hence this piece of code.

If I remove this code, the processing spinning circle will be missing.
Tell me whether you want to keep as is or whether you propose another solution.

jkppr replied:
Good catch, thanks for the explanation. In that case, let's keep it as is.

jbaptperez commented on timesketch/frontend-ng/src/components/Analyzer/TimelineSearch.vue:
Same thing here, removing the code would prevent selecting a processing timeline when creating a new visualization.
According to your approbation above, I keep the code.

jbaptperez commented:
@jkppr, all the requested modifications are done.
We also introduce a new change: `TimelineTable.vue`, function `timelineStyle(timelineStatus, isSelected)`.
Now, when a processing timeline is disabled, it is grayed out not only in the search page but also in the left menu, for consistency.
![image](https://github.com/user-attachments/assets/fc7a12c9-01ac-4f14-bf9e-e1bf0fc26c9e)

jkppr reviewed (approved):
Thanks for your contribution.
